relays/equivocation/src/equivocation_loop.rs

```diff
@@ -0,0 +1,327 @@
+				vec![]
+			},
+		}
+	}
+
+	async fn report_equivocation(&mut self, at: P::Hash, equivocation: P::EquivocationProof) {
+		match self.reporter.submit_report(&self.source_client, at, equivocation.clone()).await {
+			Ok(_) => {},
+			Err(e) => {
+				log::error!(
+					target: "bridge",
+					"Could not submit equivocation report to {} for {equivocation:?}: {e:?}",
+					P::SOURCE_NAME,
+				);
+
+				// Reconnect source client and move on
+				self.reconnect_source_client(e).await;
+			},
+		}
+	}
+
+	async fn check_block(
+		&mut self,
+		block_num: P::TargetNumber,
+		context: &mut EquivocationReportingContext<P>,
+	) {
+		let synced_headers = self.synced_source_headers_at_target(block_num).await;
+
+		for synced_header in synced_headers {
+			self.finality_proofs_buf.fill(&mut self.finality_proofs_stream);
+
+			let equivocations = match P::EquivocationsFinder::find_equivocations(
+				&context.synced_verification_context,
+				&synced_header.finality_proof,
+				self.finality_proofs_buf.buf().as_slice(),
+			) {
+				Ok(equivocations) => equivocations,
+				Err(e) => {
+					log::error!(
+						target: "bridge",
+						"Could not search for equivocations in the finality proof \
+						for source header {:?} synced at target block {block_num:?}: {e:?}",
+						synced_header.finality_proof.target_header_hash()
+					);
+					continue
+				},
+			};
+			for equivocation in equivocations {
+				self.report_equivocation(context.synced_header_hash, equivocation).await;
+			}
+
+			self.finality_proofs_buf
+				.prune(synced_header.finality_proof.target_header_number(), None);
+			context.update(synced_header);
+		}
+	}
+
+	async fn run(&mut self, tick: Duration, exit_signal: impl Future<Output = ()>) {
+		let exit_signal = exit_signal.fuse();
+		futures::pin_mut!(exit_signal);
+
+		loop {
+			// Make sure that we are connected to the source finality proofs stream.
+			match self.finality_proofs_stream.ensure_stream(&self.source_client).await {
+				Ok(_) => {},
+				Err(e) => {
+					log::error!(
+						target: "bridge",
+						"Could not connect to the {} `FinalityProofsStream`: {e:?}",
+						P::SOURCE_NAME,
+					);
+
+					// Reconnect to the source client if needed
+					match self.reconnect_source_client(e).await {
+						true => {
+							// Connection error. Move on.
+							continue
+						},
+						false => {
+							// Irrecoverable error. End the loop.
+							return
+						},
+					}
+				},
+			}
+			// Check the status of the pending equivocation reports
+			self.reporter.process_pending_reports().await;
+
+			// Check the next block
+			self.update_until_block_num().await;
+			if self.from_block_num <= self.until_block_num {
+				let mut context = match self.build_context(self.from_block_num).await {
+					Some(context) => context,
+					None => return,
+				};
+
+				self.check_block(self.from_block_num, &mut context).await;
+
+				self.from_block_num = self.from_block_num.saturating_add(1.into());
+			}
```

Comment:
This could be a loop `while self.from_block_num <= self.until_block_num {}` and do full catch-up checks in a single tick.

Doesn't look like that much work to have latency concerns..

Also I would actually suggest to skip the checking full (sub)chain and just check latest (source highest/best finalized as seen by target) - only if that has different hash than block on source at same height, then you know it forked somewhere since last tick and you can iterate to find the fork. WDYT?

Reply:
> I think an attacker(s) may be able to submit `{ submit_finality_proof(forked_header#100), submit_message(malicious_message), submit_finality_proof(good_header#101) }` in a single block. So if you'll just look at the latest, you may skip this forked_header submission. So imo we should look at all headers here

but how would `good_header#101` have the correct hash if it's been built on top of `bad_header#100`? Once relayer introduces any forked header, all subsequent child chain will be a fork, no?

Reply:
We don't check the connection between headers anywhere. So it could be built on top of other "normal" header. E.g. in following schema:
```
99 ---> 100 ---> 101
     \--> 100' 
```
the relayer1 may track the bad fork and submit the `100'` first and relayer2 may then submit the `101`

Reply:
> We don't check the connection between headers anywhere.

oh, ok, I thought (without checking the code) we did.

valid scenario then, let's keep the equivocation check for every submitted header 👍